Video streaming updates (#413)
smartdevicelink/sdl_app_library_guides: docs/Video Streaming for Navigation Apps/Video Streaming - Java/index.md

```

### Handling HMI Scaling (RPC v6.0+)

> If the HMI scales the video stream, you will have to handle scaling the projected view, touches and haptic rectangles yourself (this is all handled for you behind the scenes in the `VideoStreamManager`  API). To find out if the HMI scales the video stream, you must for query and check the `VideoStreamingCapability` for the `scale` property. Please check the [Adaptive Interface Capabilities](Displaying a User Interface/Adaptive Interface Capabilities) section for more information on how to query for this property using the system capability manager.

Reviewer: ```suggestion
If the HMI scales the video stream, you will have to handle scaling the projected view, touches and haptic rectangles yourself (this is all handled for you behind the scenes in the `VideoStreamManager` API). To find out if the HMI scales the video stream, you must for query and check the `VideoStreamingCapability` for the `scale` property. Please check the [Adaptive Interface Capabilities](Displaying a User Interface/Adaptive Interface Capabilities) section for more information on how to query for this property using the system capability manager.
```

### Video Streaming Parameters (SDL v5.1+)
Starting with SDL version 5.1+ the `VideoStreamingParameters` you provide will automatically be aligned with the `VideoStreamingCapabilities` provided by the HMI.

> If the HMI provides the scale or resolution in the `VideoStreamingCapabilities` the Video stream will use that scale or resolution. Otherwise, the scale or resolution defined in the `VideoStreamingParameters` you provided will be used.

Reviewer: ```suggestion
If the HMI provides the scale or resolution in the `VideoStreamingCapabilities` the video stream will use that scale or resolution. Otherwise, the scale or resolution you defined in the `VideoStreamingParameters` will be used.
```

> If the HMI provides the BitRate or preferred frame rate in the `VideoStreamingCapabilities` and they are also defined in the `VideoStreamingParamerters` you provided the smaller BitRate or preferred frame rate will be used.

Reviewer: ```suggestion
If the HMI provides the bitrate or preferred frame rate in the `VideoStreamingCapabilities` and they are also defined in the `VideoStreamingParamerters` you provided, the smaller bitrate or preferred frame rate will be used.
```

> ### Video FrameRate (RPC v7.1+)

Reviewer: ```suggestion
### Video Framerate
```

> Starting with RPC version 7.1+ you can define a frame rate within your `VideoStreamingParameters` that you would like your video to stream at. You can also have the option to turn this capability on or off by using the `stableFrameRate` flag in the `VideoStreamingParameters`.

Reviewer: ```suggestion
Starting with SDL v5.1, the video stream manager works behind the scene to create a consistent video stream that matches the framerate set in `VideoStreamingParameters`. This is the now the default behavior but you have the option to revert to the old behavior by setting the `stableFrameRate` flag to `false` in the `VideoStreamingParameters`.
```

```java
VideoStreamingParameters params = new VideoStreamingParameters();
//Turn on use of stable frame rate
params.setStableFrameRate(true);
//Set the frame rate that you would wish to stream at
params.setFrameRate(30);
```

### Supporting Different Video Streaming Window Sizes (RPC v7.1+)

> You can specify two `VideoStreamingRange` parameters when you want to start your video stream using the `startRemoteDisplay` method, one range will be for landscape orientation and one range will be for portrait orientation.

Reviewer: ```suggestion
Some HMIs support multiple view sizes and may resize your SDL app's view during video streaming (i.e. to a collapsed view, split screen, preview mode or picture-in-picture). By default, your app will support all the view sizes and the `VideoStreamManager` will resize the video stream when the HMI notifies the app of the updated screen size. 
If you you wish to support only some screen sizes, you can configure the two `VideoStreamingRange` parameters when starting your video stream using the `startRemoteDisplay` method. One range is for landscape orientations and one range is for portrait orientations.
```

> In these `VideoStreamingRange` parameters you can define different view sizes that you wish to support in the event that the HMI resizes the view during the stream. (i.e. to a collapsed view, split screen, preview mode or picture-in-picture)

Reviewer: ```suggestion
In these `VideoStreamingRange` parameters you can define different view sizes that you wish to support in the event that the HMI resizes the view during the stream.
```

> In the `VideoStreamingRange` you will define a minimum and maximum Resolution, minimum diagonal, and a minimum and maximum aspect Ratio. Any values you do not wish to use should be set to `null`.

Reviewer: ```suggestion
In the `VideoStreamingRange` you will define a minimum and maximum resolution, minimum diagonal, and a minimum and maximum aspect ratio. Any values you do not wish to use should be set to `null`.
```

final VideoStreamingRange disabledRange = new VideoStreamingRange(new Resolution(0, 0), new Resolution(0, 0), 0.0, 0.0, 0.0);

//This VideoStreamingRange represents that we will support any resolution between 500x200 and 800x400 no matter the diagonal size or aspect ratio
final VideoStreamingRange landscapeRange = new VideoStreamingRange(new Resolution(500, 200), new Resolution(800, 400), null, null, null);

//This VideoStreamingRange represents that we will support any aspect ratio between 1.0 and 2.5 no matter the resolution or diagonal size
final VideoStreamingRange portraitRange = new VideoStreamingRange(null, null, null, 1.0, 2.5);

if (sdlManager.getVideoStreamManager() != null) {
    sdlManager.getVideoStreamManager().start(new CompletionListener() {
        @Override
        public void onComplete(boolean success) {
            if (success) {
                sdlManager.getVideoStreamManager().startRemoteDisplayStream(getApplicationContext(), MyDisplay.class, null, false, landscapeRange, portraitRange);
            } else {
                DebugTool.logError(TAG, "Failed to start video streaming manager");
            }
        }
    });
}
```

!!! NOTE

> If you disable both the `supportedLandscapeStreamingRange` and `supportedPortraitStreamingRange`, the video will not stream

Reviewer: ```suggestion
If you disable both the `supportedLandscapeStreamingRange` and `supportedPortraitStreamingRange`, video will not stream.
```

!!!

> If the HMI resizes the view during the stream, the video stream will automatically restart with the new size and the `onViewResized` method you defined in your presentation class.

Reviewer: ```suggestion
If the HMI resizes the view during the stream, the video stream will automatically restart with the new size and the `onViewResized` method you defined in your presentation class will be notified of the new screen size.
```

```

### Handling HMI Scaling (RPC v6.0+)

> If the HMI scales the video stream, you will have to handle scaling the projected view, touches and haptic rectangles yourself (this is all handled for you behind the scenes in the `VideoStreamManager` API). To find out if the HMI scales the video stream, you must for query and check the `VideoStreamingCapability` for the `scale` property. Please check the [Adaptive Interface Capabilities](Displaying a User Interface/Adaptive Interface Capabilities) section for more information on how to query for this property using the system capability manager.

Reviewer: I wonder if "Handling HMI Scaling" should be moved to the bottom of the guide since this paragraph is only relevant if the developer isn't using the Video Stream Manager.
